google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java

```diff
@@ -568,28 +583,84 @@ public void run() {
    * next group of ready batches.
    */
   private void sendBatch(final BulkCommitBatch batch) {
-    boolean success = rateLimiter.tryMakeRequest(batch.getOperationCount());
+    boolean success = rateLimiter.tryMakeRequest(batch.getPendingOperationCount());
     Preconditions.checkState(success, "Batch should be under rate limit to be sent.");
-    try {
-      final ApiFuture<List<BatchWriteResult>> commitFuture = batch.bulkCommit();
-      commitFuture.addListener(
-          new Runnable() {
-            public void run() {
-              try {
-                batch.processResults(commitFuture.get(), null);
-              } catch (Exception e) {
-                batch.processResults(new ArrayList<BatchWriteResult>(), e);
+    MoreExecutors.directExecutor()
+        .execute(
+            new Runnable() {
+              public void run() {
+                bulkCommit(batch);
+                boolean removed = batchQueue.remove(batch);
+                Preconditions.checkState(removed, "The batch should be in the BatchQueue.");
+                sendReadyBatches();
               }
-              // Remove the batch from BatchQueue after it has been processed.
-              boolean removed = batchQueue.remove(batch);
-              Preconditions.checkState(removed, "The batch should be in the BatchQueue.");
-              sendReadyBatches();
-            }
-          },
-          MoreExecutors.directExecutor());
-    } catch (Exception e) {
-      batch.processResults(new ArrayList<BatchWriteResult>(), e);
+            });
+  }
+
+  private void bulkCommit(BulkCommitBatch batch) {
```

Comment:
What do you think about having the retry logic live in `BulkWriter` instead of inside `UpdateBuilder`? I originally wanted to take care of everything in `UpdateBuilder`, but then I realized that doing so would prohibit creating a new `BulkCommitBatch` via constructor.

Reply:
This matches Node so I am fully supportive.